Document filtering by type for children endpoint

jonhealy1 opened the pull request:
This PR proposes adding an optional `type` query parameter to the `/children` endpoint to support filtering results by resource type (e.g., `Catalog` or `Collection`).

### Motivation

While the specification recommends separating different types into different hierarchy branches as a best practice, real-world data organization often results in "mixed content" folders (e.g., a Project Catalog containing both sub-project Catalogs and direct Dataset Collections).

In dynamic, database-backed APIs (like those using Elasticsearch), retrieving and paginating mixed-type lists can be computationally expensive and complex to implement.

Adding a standard `type` parameter allows:
1.  **Backend Optimization:** Implementations can execute efficient, type-specific queries (e.g., `type: "Catalog"`) rather than fetching all children and filtering in memory.
2.  **Client Flexibility:** Clients can request only the resource type they are interested in (e.g., "Show me sub-folders" vs. "Show me datasets") without receiving unwanted data.
3.  **Standardization:** Prevents the proliferation of custom filter parameters (e.g., `?kind=`, `?filter=`) across different implementations.

### Proposed Change

Add a section detailing the `type` parameter:

#### Filtering by Type
Implementations MAY support a `type` query parameter to allow clients to request a specific resource type.

* `GET .../children?type=Catalog`
* `GET .../children?type=Collection`

**Related Issue(s):** 

- #

**Proposed Changes:**

1. 
2. 

**PR Checklist:**

- [ ] This PR has **no** breaking changes.
- [ ] I have added my changes to the [CHANGELOG](https://github.com/stac-api-extensions/children/blob/main/CHANGELOG.md) **or** a CHANGELOG entry is not required.

m-mohr committed:
Generalize /children endpoints to be usable outside of landing page #3, general rewrite/update #2

jonhealy1 committed:
Document filtering by type for children endpoint

This PR proposes adding an optional `type` query parameter to the `/children` endpoint to support filtering results by resource type (e.g., `Catalog` or `Collection`).

### Motivation

While the specification recommends separating different types into different hierarchy branches as a best practice, real-world data organization often results in "mixed content" folders (e.g., a Project Catalog containing both sub-project Catalogs and direct Dataset Collections).

In dynamic, database-backed APIs (like those using Elasticsearch), retrieving and paginating mixed-type lists can be computationally expensive and complex to implement.

Adding a standard `type` parameter allows:
1.  **Backend Optimization:** Implementations can execute efficient, type-specific queries (e.g., `type: "Catalog"`) rather than fetching all children and filtering in memory.
2.  **Client Flexibility:** Clients can request only the resource type they are interested in (e.g., "Show me sub-folders" vs. "Show me datasets") without receiving unwanted data.
3.  **Standardization:** Prevents the proliferation of custom filter parameters (e.g., `?kind=`, `?filter=`) across different implementations.

### Proposed Change

Add a section detailing the `type` parameter:

#### Filtering by Type
Implementations MAY support a `type` query parameter to allow clients to request a specific resource type.

* `GET .../children?type=Catalog`
* `GET .../children?type=Collection`

m-mohr reviewed (changes requested):
I think I'm fine with the functionality in general, but as it's optional clients may run into issues as they don't know whether it's supported or not. As such, I'd propose to make it an optional conformance class. Something like <https://api.stacspec.org/v1.0.0-rc.2/children#type-filter>. Should be listed at the top and in the section.

Additionally, you'd need to update the TOC and the OpenAPI document.

m-mohr commented on README.md:
```diff
@@ -81,6 +81,15 @@ implementations may only return a single type if the children only consist of a
 
+### Filtering by Type
+
+Because the `children` array is polymorphic (containing both `Catalog` and `Collection` objects), implementations MAY support a `type` query parameter to allow clients to filter the response to a specific resource type.
```
I think this should make a little more clear that servers are filtering and not e.g. converting Collections to Catalogs when ?type=Catalog is requested.

```suggestion
Because the `children` array is polymorphic (containing both `Catalog` and `Collection` objects),
implementations MAY support a `type` query parameter to allow clients to filter the response to a specific resource type.
Results SHALL be *filtered*, a conversion between Catalog and Collection is not foreseen.
```

jonhealy1 committed:
Update README.md

Co-authored-by: Matthias Mohr <m.mohr@moregeo.it>

m-mohr commented on README.md:
```diff
@@ -81,6 +81,15 @@ implementations may only return a single type if the children only consist of a
+* `GET .../children?type=Catalog` - Returns only child Catalogs.
+* `GET .../children?type=Collection` - Returns only child Collections.
+
+This is recommended for implementations where backend storage (e.g., Elasticsearch indices) or client logic benefits from strict typing.
```
I don't really understand how this helps in the backend implementation as people can still choose not to provide the type query parameter. Can you elaborate on this a bit? Doesn't necessarily need to be in the text, just curious.

jonhealy1 replied:
True, but it would be difficult to combine an items search on both collections and catalogs as they may live in different db indexes.

jonhealy1 replied:
The comment can be removed if it's just confusing.

m-mohr replied:
This can only contain catalogs and collections (rel=child), items can't be part of the response as they would be rel=item.
I'm fine to keep it if it's still valid.

jonhealy1 replied:
I opened an issue about allowing items search on the /children endpoint #7

m-mohr replied:
Responded there, resolving here.

m-mohr commented on README.md:
```diff
@@ -81,6 +81,17 @@ implementations may only return a single type if the children only consist of a
 It is considered a best practice to structure the hierarchy in a way that the children for each 
 individual request only consist of a single type.
 
+### Filtering by Type
```
```suggestion
### Filtering by Type

This section describes an OPTIONAL conformance class that adds a `type` query parameter for filtering:

- **Conformance Class:** <https://api.stacspec.org/v1.0.0-rc.2/children#type-filter>
```

jonhealy1 committed:
Add conformance class for type filtering

Co-authored-by: Matthias Mohr <m.mohr@moregeo.it>

m-mohr commented on README.md:
```diff
@@ -81,6 +81,21 @@ implementations may only return a single type if the children only consist of a
+
+This section describes an OPTIONAL conformance class that adds a `type` query parameter for filtering:
+
+- **Conformance Classes:** <https://api.stacspec.org/v1.0.0-rc.2/children#type-filter>
```
```suggestion
- **Conformance Class:** <https://api.stacspec.org/v1.0.0-rc.2/children#type-filter>
```

m-mohr committed:
Update README.md

m-mohr commented:
> Additionally, you'd need to update the TOC and the OpenAPI document.

@jonhealy1 I think this^ is still open. Now that #6 is merged, the CI also fails due to the outdated TOC, I think. Can you please update this PR accordingly? Otherwise, I think it looks good.

Changed the base after the merging #6 and tried to resolve the conflicts...

m-mohr committed:
Merge branch 'main' into patch-2

jonhealy1 committed:
Fix Markdown header for Filtering by Type section

Updated the section header for 'Filtering by Type' to use proper Markdown formatting.

jonhealy1 committed:
Add 'type' query parameter to getChildren operation

Added 'type' query parameter to filter catalogs or collections.

jonhealy1 committed:
Fix syntax

jonhealy1 committed:
Refactor parameters section in openapi.yaml

Formatted parameters section for consistency.

jonhealy1 committed:
Fix indentation in openapi.yaml parameters section

jonhealy1 commented:
@m-mohr Passing :)

m-mohr commented on README.md:
```diff
@@ -233,6 +266,11 @@ The `GET /children` endpoint response object could look as follows:
+    {
+      "rel": "parent",
+      "type": "application/json",
+      "href": "https://stac-api.example.com"
```
```suggestion
      "href": "https://stac-api.example"
```

m-mohr replied:
I'm also not sure why this is added here? Might come from an old version?

m-mohr commented on README.md:
```diff
@@ -214,6 +230,23 @@ The `GET /children` endpoint response object could look as follows:
+          "type": "application/json",
+          "title": "Flight 2",
+          "href": "https://stac-api.example/drones/flight-2"
+        },
```
The links seem to be duplicate.

jonhealy1 committed:
Update README.md

Co-authored-by: Matthias Mohr <m.mohr@moregeo.it>

jonhealy1 committed:
Fix typos and remove redundant links in README

m-mohr reviewed (commented):
Thanks

Edit: Approving via phone on GH doesn't work, it seems. Will do once back at my desk.

jonhealy1 commented:
@m-mohr Thanks :)
